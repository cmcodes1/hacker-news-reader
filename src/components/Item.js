import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import dateFromNow from '../helpers/DateFromNow';
import colors from '../theme/colors';

export default function Item({ index, title, url, score, by, time, descendants, }) {

    const getHostname = (url) => {
        const matches = url.match(/^https?\:\/\/([^\/?#]+)(?:[\/?#]|$)/i);
        return matches && matches[1];
    }

    return (
        <View style={styles.item}>
            <Text style={styles.heading}>
                <Text style={styles.bigText}>{index + 1}{'. \u25B2 '}</Text>
                <Text style={styles.title}>{title}</Text>
                <Text style={styles.bigText}>{` (${getHostname(url)})`}</Text>
            </Text>
            <Text style={styles.smallText}>{score}{' points by '}{by}{' '}{dateFromNow(time)}{descendants && ` | ${descendants} comments`}</Text>
        </View>
    )
}

const styles = StyleSheet.create({
    item: {
        margin: 10,
    },
    heading: {
        flexDirection: 'row',
        flexWrap: 'wrap',
    },
    title: {
        fontSize: 14,
        color: colors.textPrimary,
    },
    bigText: {
        fontSize: 14,
    },
    smallText: {
        fontSize: 11,
    },
})