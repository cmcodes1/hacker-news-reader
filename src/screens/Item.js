import React, { useState, useEffect, } from 'react'
import { View, Text, StyleSheet, Image, ActivityIndicator, FlatList, } from 'react-native'
import CommentsService from '../api/CommentsService'
import yCombinatorLogo from '../assets/yCombinatorLogo.png'
import Comment from '../components/Comment'
import colors from '../theme/colors'

export default function Item({ navigation, route }) {

    const commentsService = new CommentsService()

    const [isLoading, setIsLoading] = useState(false)

    const [comments, setComments] = useState([])

    useEffect(() => {
        fetchComments()
    }, [navigation])

    const fetchComments = async () => {

        setIsLoading(true)

        let promises = []

        let commentsData = []

        route.params.kids.forEach((commentId) => {
            promises.push(commentsService.getComments(commentId));
        })

        commentsData = await Promise.all(promises)

        setComments(commentsData)

        setIsLoading(false)

    }

    // useEffect(() => {
    //     console.log('comments', comments)
    // }, [comments])

    const renderComment = ({ item }) => (
        <Comment
            navigation={navigation}
            by={item.by}
            kids={item.kids}
            text={item.text}
            time={item.time}
        />
    )

    return (
        <View style={styles.container}>
            <View style={styles.header}>
                <Image style={styles.logo} source={yCombinatorLogo} />
                <Text style={styles.headerText}>Hacker News</Text>
            </View>
            <View style={styles.body}>
                {
                    isLoading
                        ?
                        <ActivityIndicator size="large" color={colors.primary} />
                        :
                        <FlatList
                            data={comments}
                            renderItem={renderComment}
                            keyExtractor={item => item.id}
                        />
                }
            </View>
        </View>
    )
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: colors.secondary,
    },
    header: {
        backgroundColor: colors.primary,
        height: 50,
        flexDirection: 'row',
        justifyContent: 'flex-start',
        alignItems: 'center',
    },
    logo: {
        height: 40,
        width: 40,
        borderWidth: 1,
        borderColor: colors.tertiary,
        marginHorizontal: 10,
    },
    headerText: {
        color: colors.textPrimary,
        fontSize: 30,
        fontWeight: 'bold',
    },
    body: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
    },
})