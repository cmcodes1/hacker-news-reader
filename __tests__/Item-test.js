import React from 'react'
import { create } from 'react-test-renderer'
import Item from '../src/screens/Item'

jest.useFakeTimers()

const tree = create(<Item />)

test('snapshot', () => {
    expect(tree).toMatchSnapshot()
})