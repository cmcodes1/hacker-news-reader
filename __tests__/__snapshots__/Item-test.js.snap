// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F6F6EF",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FF6600",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "flex-start",
      }
    }
  >
    <Image
      source={
        Object {
          "process": [Function],
        }
      }
      style={
        Object {
          "borderColor": "#FFFFFF",
          "borderWidth": 1,
          "height": 40,
          "marginHorizontal": 10,
          "width": 40,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 30,
          "fontWeight": "bold",
        }
      }
    >
      Hacker News
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <RCTScrollView
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      viewabilityConfigCallbackPairs={Array []}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;
